implementation/src/main/java/io/smallrye/config/KeyMapBackedConfigSource.java

```diff
@@ -0,0 +1,38 @@
+package io.smallrye.config;
+
+import java.util.Map;
+
+import org.eclipse.microprofile.config.spi.ConfigSource;
+
+import io.smallrye.config.common.AbstractConfigSource;
+
+public abstract class KeyMapBackedConfigSource extends AbstractConfigSource {
+    private static final long serialVersionUID = 4378754290346888762L;
+
+    private final KeyMap<String> properties;
+
+    public KeyMapBackedConfigSource(final String name, final KeyMap<String> properties) {
+        super(name, ConfigSource.DEFAULT_ORDINAL);
+        this.properties = properties;
+    }
+
+    public KeyMapBackedConfigSource(final String name, final int ordinal, final KeyMap<String> properties) {
+        super(name, ordinal);
+        this.properties = properties;
+    }
+
+    @Override
+    public Map<String, String> getProperties() {
+        return properties.toMap();
+    }
+
+    @Override
+    public String getValue(final String propertyName) {
+        final String value = properties.findRootValue(propertyName);
+        if (value == null) {
+            // In case of searching for an empty root.
```

Comment:
An empty root should not mean an empty initial segment, otherwise at a minimum we're creating a useless extra map.  There's nothing special about "root" values from the perspective of a key map, so if values are getting added like `".foo"` then the problem should be addressed on the add side, not this side.